File: leetcode/27.remove-element/cpp/removeelement_test.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"

#include "removeelement.cpp"

struct testCase
{
    std::vector<int> nums;
    int val;
    int want;
    std::vector<int> wantNums;
};

TEST_CASE("RemoveElement")
{
    std::vector<testCase> tcs{
        {std::vector<int>{3, 2, 2, 3}, 3, 2, std::vector<int>{2, 2}},
        {std::vector<int>{0, 1, 2, 2, 3, 0, 4, 2}, 2, 5, std::vector<int>{0, 1, 3, 0, 4}},
        {std::vector<int>{}, 2, 0, std::vector<int>{}},
        {std::vector<int>{0, 0, 0}, 0, 0, std::vector<int>{}},
    };

    Solution s;

    for (auto tc : tcs)
    {
        int got = s.removeElement(tc.nums, tc.val);
        CHECK(got == tc.want);
        std::vector<int> gotNums(tc.nums.begin(), tc.nums.begin() + got);
        CHECK(gotNums == tc.wantNums);
    }
}

File: leetcode/27.remove-element/ts/removeelement.test.ts
import { expect, test } from "vitest";
import removeElement from "./removeelement.js";

interface testCase {
    nums: number[],
    val: number,
    want: number,
    wantNums: number[]
}

test("removeelement", () => {
    const tcs: testCase[] = [
        { nums: [3, 2, 2, 3], val: 3, want: 2, wantNums: [2, 2] },
        { nums: [0, 1, 2, 2, 3, 0, 4, 2], val: 2, want: 5, wantNums: [0, 1, 3, 0, 4] },
        { nums: [], val: 2, want: 0, wantNums: [] },
        { nums: [0, 0, 0], val: 0, want: 0, wantNums: [] }
    ]

    for (const tc of tcs) {
        const got = removeElement(tc.nums, tc.val)
        expect(got).toEqual(tc.want)
        expect(tc.nums.slice(0, got)).toEqual(tc.wantNums)
    }
})
